### PR TITLE
Advance the clock if it's before the build timestamp

### PR DESCRIPTION
**Pull request by fhunleth**
This brings us out of the 1970s and more importantly, makes some SSL
certs work without waiting on NTP.

### CODE REVIEW

File: src/erlinit.c [hunk 1]
```diff
@@ -490,6 +491,31 @@ static void setup_environment(const struct erl_run_info *run_info)
...
+    if (clock_gettime(CLOCK_REALTIME, &tp) < 0) {
+        warn("clock_gettime failed. Skipping time check");
+        return;
+    }
+
+    if (tp.tv_sec < BUILD_TIME) {
+        tp.tv_sec = BUILD_TIME;
+        tp.tv_nsec = 0;
+#ifdef UNITTEST
+        fatal("How can the current time be before the build time when unittesting?");
```
Comment (electricshaman) [inline src/erlinit.c]: unit testing*

**Comment by ConnorRigby** [comment]
For what it's worth, The `nerves_system_XXX` will need updating to get this change. It might be worth putting in a one time warning about this change.

**Comment by electricshaman** [comment]
@fhunleth Along the same lines, how do you feel about `fake-hwclock`?

**Comment by fhunleth** [comment]
`fake-hwclock` is cool. It seems reasonable to make an Elixir module to do the same thing that it does should the need arise.

**Comment by GregMefford** [comment]
I think this will help with making certificates' `notBefore` field validation work better, but isn't a real fix for the problem. For example, Google rolls their certificates every 3 months, so you'd have to be sure to deploy a new firmware pretty frequently to be able to rely on this. Also, CRL checks could cause failures because they tend to have tighter time-windows than certificate issuance.

screen shot 2018-02-12 at 10 20 57 am

I'm not saying we shouldn't do this, but we should think about the trade-off between being able to easily hard-code a check that it's not 1970 against the benefit of working some (but not all) of the time. If your device is intended to be Internet-connected, maybe you _should_ wait for your clock to sync before validating certificates, and we just need better support for reliable NTP at boot time.

**Comment by fhunleth** [comment]
I have some heartburn supporting the "check if the year is 1970 to see whether NTP is synchronized" argument. That should be fixed regardless of this issue.

**Comment by GregMefford** [comment]
Well what if we had an API that just keeps track of whether NTP has successfully synced the time since boot, e.g. using a GenServer to manage `ntpd`? That's the kind of thing I had in mind when I mentioned "better support for reliable NTP at boot time."

We could hook it into `SystemRegistry` so that people could block on it if they want to, like they might for having the IP address set. I think we shouldn't/can't block the application start-up process because there are likely to be lots of cases where you want the time to be set, but you can safely continue doing other things until it's set.

**Comment by electricshaman** [comment]
We should absolutely block until the time is set.

**Comment by entone** [comment]
Blocking until time is set should not be the default, but an option for
sure.

**Comment by fhunleth** [comment]
@GregMefford @electricshaman Please open an issue on https://hex.pm/packages/nerves_ntp to address the need to tell whether NTP has synchronized or not. That would be a better place for this discussion.

**Comment by ConnorRigby** [comment]
I like the idea of having an option for blocking until time is set. It certainly can't be default in case you have an app that doesn't require internet.

**Comment by fhunleth** [comment]
FYI - I plan on revising this PR so that the time update is optional and it will default to not updating the time.

**Pull request merged**